cli/src/main/java/org/eclipse/hono/cli/client/ClientConfig.java

```diff
@@ -0,0 +1,95 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.cli.client;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.eclipse.hono.config.ClientConfigProperties;
+import org.springframework.core.MethodParameter;
+import org.springframework.shell.CompletionContext;
+import org.springframework.shell.CompletionProposal;
+import org.springframework.shell.standard.ValueProviderSupport;
+import org.springframework.stereotype.Component;
+
+/**
+ * Configuration class. Holds all the parameters to execute a command.
+ * <p>
+ * A default instance will be created at startup and it will be cloned and modified with custom
+ * parameter for the execution of the command.
+ */
+public class ClientConfig implements Cloneable {
+    public static final String TYPE_TELEMETRY = "telemetry";
+    public static final String TYPE_EVENT = "event";
+    public static final String TYPE_ALL = "all";
+
+    public ClientConfigProperties honoClientConfig;
+
+    public String tenantId;
+    public String deviceId;
+    public String messageType;
+    public int connectionRetryInterval;
+    public int requestTimeoutInSecs;
+    public String messageAddress;
+    public String payload;
+
+    /**
+     * Constructor.
+     *
+     * @param tenantId Parameter for command execution
+     * @param deviceId Parameter for command execution
+     * @param messageType Parameter for command execution
+     * @param connectionRetryInterval Parameter for command execution
+     * @param requestTimeoutInSecs Parameter for command execution
+     * @param messageAddress Parameter for command execution
+     * @param payload Parameter for command execution
+     */
+    public ClientConfig(final String tenantId, final String deviceId, final String messageType, final int connectionRetryInterval, final int requestTimeoutInSecs, final String messageAddress, final String payload) {
```

Comment:
it is kind of hard to understand why message payload would be part of client configuration. On the other hand, the host and port to connect to and the credentials seem to be missing. Or am I mistaken?

Reply:
Actually both the constructor and the name class are pretty misleading...
I'll make you a question before fix it:
I wanted to maintain the spring parameters compatibility at startup and those are the parameters in the ApplicationClient annotated with the `@Value` tag.
Those parameters constitute the basic env to execute the commands while the other parameters like host, port etc can be inserted at the moment of executing a single command 
(e.g. receive --host --port --username --password --tenant --message)
Said that, i'm not an expert of UX (maybe the opposite) but i made a thought on a possible use case:
As a user i want to send many event with the same looong payload

than a great behaviour will be:

> java -jar hono-cli-*-exec.jar --message.payload="that's a long one" --tenant.id=$MY_TENANT --device.id=$MY_DEVICE

>HONO:> send --hostname $AMQP_NETWORK_IP --port 15672 --username me --password secret
>HONO:> send --hostname $AMQP_NETWORK_IP --port 15672 --username me --password secret
>HONO:> send --hostname $AMQP_NETWORK_IP --port 15672 --username me --password secret --message "this time the message it's short"
>HONO:> send --hostname $AMQP_NETWORK_IP --port 15672 --username me --password secret

In this way you don't have to insert the credential directly in the shell but you can have the advantage of the sping env. The question is: what are the parameters that is better to acquire at startup and what in the command? And also the parameters can be acquired in a more straightforward way (but more rambling):

>java -jar hono-cli-*-exec.jar

>HONO:> send --hostname $AMQP_NETWORK_IP --port 15672 
>What message do you want to send?
>HONO:> "Here the body message"
>What is the message type? ["event","telemetry"]
>HONO:> "event"
>Message sent

Also i wanted to underline the concept that the history of the commands executed (like up and down arrow on a regular shell) it's not present.

Reply:
My understanding is that the shell will be used to do two different things
1. interact with Hono's north bound API, i.e. act as an application, and
2. interact with Hono's AMQP protocol adapter, i.e. act as a device

FMPOV the commands for doing so should therefore be put into one class each which also maintains the connection state to the corresponding endpoint, i.e. either the dispatch router or the AMQP adapter. These classes can then be annotated with `@ShellComponent` and their (command) methods with `@ShellCommand`. I also think that these classes should provide means to establish a connection and to disconnect and that other commands, e.g. for sending telemetry or events, should be available only once the connection has been established.

That said, it might make sense to be able to provide *default* values for e.g. the host and port to connect to or the username and password so that when running the *connect* command, you do not need to specify them again. This would also alleviate the user from having to specify any of the connection related properties when using e.g. *sendTelemetry* or *sendEvent*.

Separating the connection establishment from the command execution itself makes the current code very hard to read and understand. Maybe you can re-factor the classes based on my comments above. WDYT?

Reply:
I guess that give the user a client that abstract the structure of Hono and let him make less choice as possible is the objective. If i got it right you are suggesting something like:

>java -jar hono-cli.jar
>HONO:>help
>Available commands:
>`        `connect AMQP-adapter [-H --host] [-P --port] [-u --username] [-p --password]
>`        `connect messegging-network [-H --host] [-P --port] [-u --username] [-p --password]
>HONO:> connect AMQP-adapter --hostname $AMQP_NETWORK_IP --port 15672 --username me --password secret
>Connected!
>HONO:>help
>... the two commands ... or disconnect or exit

or maybe you prefer a
>java -jar hono-cli.jar --adapter --hostname $AMQP_NETWORK_IP --port 15672 --username me --password secret
>HONO:>help
>... the two commands ... or disconnect=exit

The second is almost the approach used now but with parameters more human-usable<

However i totally agree to separate the acquisition of parameters by context as you proposed.

Also i've another question: the comments and the messages along with the commands must refer to the behaviour or to the connection aspects? i.e.:
"Act as a device" or "Connect to the AMQP-adapter". The user want to use the client to test how works a device or want to test the adapter?
I speak as a newcome to Hono and if i can say my two cents:
If I don't know how all the structure works but I want to make some test,, having many terms (that yes refers to the same thing) it confuse me. At least that was my personal first approach with Hono.

Reply:
> f i got it right you are suggesting something like

exactly :-)

> Also i've another question: the comments and the messages along with the commands must refer to the behaviour or to the connection aspects? i.e.:
"Act as a device" or "Connect to the AMQP-adapter". The user want to use the client to test how works a device or want to test the adapter?

In the end, the CLI can be used to do both. New users will most likely use it to *act as a device* as part of the getting started guide, whereas the developers of adapters will probably want to use the CLI to test the adapters.

> I speak as a newcome to Hono and if i can say my two cents:
If I don't know how all the structure works but I want to make some test,, having many terms (that yes refers to the same thing) it confuse me. At least that was my personal first approach with Hono.

Totally valid point. Do you have a suggestion as to how we could improve the terminology and make it less confusing?